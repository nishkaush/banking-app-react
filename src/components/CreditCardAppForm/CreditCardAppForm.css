.credit__card__app__form {
  display: flex;
  flex-direction: column;
}

.credit__card__app__form input {
  border: 1px solid white;
  padding: 0% 1% 1% 1%;
  /* border-top: 1px solid #eee; */
  border-bottom: 1px solid #eee;
  margin: 0% 1% 1% 1%;
  padding-left: 2%;
  border-radius: 10px;
  font-size: 0.9em;
}

.credit__card__app__form input:focus {
  outline: none;
  /* background: #eee */
}

.credit__card__app__form label {
  font-size: 0.7em;
  padding-left: 3%;
  padding-bottom: 0.5%;
  color: #939393;
  margin-top: 1.5%;
}

.credit__card__app__form .btn_container {
  width: 60%;
  display: flex;
  justify-content: space-around;
  margin: 2% auto;
}

.credit__card__app__form button {
  padding: 3% 4%;
  background: royalblue;
  color: white;
  width: 50%;
  font-size: 0.9em;
  border: none;
  margin: 2%;
  border-radius: 5px;
  cursor: pointer;
  box-shadow: 0px 10px 30px rgba(0, 0, 0, 0.3);
}

.credit__card__app__form button:hover {
  background: rgb(94, 123, 211);
}

input[type="number"]::-webkit-outer-spin-button,
input[type="number"]::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

input[type="number"] {
  -moz-appearance: textfield;
}

input:focus::placeholder {
  opacity: 0;
}


.error__msg__helper {
  padding-left: 3%;
  font-size: 0.6em;
  margin-bottom: 1%;
}

.error__msg__helper span {
  /* margin: 0 0.5%; */
  color: red;
}