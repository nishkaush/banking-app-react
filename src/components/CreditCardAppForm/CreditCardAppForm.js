import React, { Component } from "react";
import { Container } from "@material-ui/core";
import { connect } from "react-redux";
import "./CreditCardAppForm.css";

class CreditCardAppForm extends Component {
  state = {
    textfieldsArr: [
      {
        type: "text",
        name: "fullName",
        label: "Full Name",
        value: "",
        touched: false,
        error: false,
        errorMessages: [],
        validation: {
          onRequired: val => (val ? true : "Full Name is required"),
          onMinLength: val =>
            val.length >= 8 ? true : "Must be atleast 8 characters long"
        }
      },
      {
        type: "email",
        name: "email",
        label: "Email",
        value: "",
        touched: false,
        error: false,
        errorMessages: [],
        validation: {
          onRequired: val => (val ? true : "Email is required"),
          onMinLength: val =>
            /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(val) === true
              ? true
              : "Must be a valid email address"
        }
      },
      {
        type: "number",
        name: "creditLimit",
        label: "Credit Limit ($)",
        value: "",
        touched: false,
        error: false,
        errorMessages: [],
        validation: {
          onRequired: val => (val ? true : "Credit Limit is required"),
          onMinAmount: val => (val > 0 ? true : "Must be greater than 0")
        }
      },
      {
        type: "number",
        name: "assetsValue",
        label: "Total Assets Value ($)",
        value: "",
        touched: false,
        error: false,
        errorMessages: [],
        validation: {
          onRequired: val => (val ? true : "Assets Value is required"),
          onMinAmount: val =>
            val >= 0 ? true : "Must be greater than or equal to 0"
        }
      },
      {
        type: "number",
        name: "liabilitiesValue",
        label: "Total Liabilities Value ($)",
        value: "",
        touched: false,
        error: false,
        errorMessages: [],
        validation: {
          onRequired: val => (val ? true : "Liabilities Value is required"),
          onMinAmount: val =>
            val >= 0 ? true : "Must be greater than or equal to 0"
        }
      }
    ]
  };

  handleOnChange(ind, event) {
    let foundObj = { ...this.state.textfieldsArr[ind] };
    foundObj.value = event.target.value;
    foundObj.touched = true;
    this.setState(state => {
      state.textfieldsArr[ind] = foundObj;
      return state;
    });
  }

  validateField(ind, event) {
    console.log("ind", ind);
    let foundObj = { ...this.state.textfieldsArr[ind] };
    if (
      foundObj.touched === true &&
      Object.entries(foundObj.validation).length
    ) {
      let flag = this.handleValidationFunctions(
        foundObj.validation,
        foundObj,
        event
      );
      this.setupErrorPropsForField(foundObj, ind, flag);
    }
  }

  handleValidationFunctions(validationObj, foundObj, event) {
    let flag = true;
    foundObj.errorMessages = [];
    for (let key in validationObj) {
      let result = validationObj[key](event.target.value);
      if (result !== true) {
        flag = false;
        foundObj.errorMessages.push(result);
      }
    }
    return flag;
  }

  setupErrorPropsForField(foundObj, ind, flag) {
    if (!flag) {
      foundObj.error = true;
      this.setState(state => {
        state.textfieldsArr[ind] = foundObj;
        return state;
      });
    } else {
      foundObj.error = false;
      foundObj.errorMessages = [];
      this.setState(state => {
        state.textfieldsArr[ind] = foundObj;
        return state;
      });
    }
  }

  handleFormSubmit(event) {
    event.preventDefault();
    let allFieldAreValid = this.checkFormValidity();
    let payload = this.preparePayloadForFormSubmission();
    allFieldAreValid && this.checkAssetsAndLiabilities()
      ? this.props.onFormSubmit(payload)
      : alert(
          "Fill out all fields first And make sure your assets are bigger than your liabilities "
        );
  }

  checkFormValidity() {
    return this.state.textfieldsArr.every(
      elm =>
        (!elm.touched && !elm.validation.onRequired) ||
        (!elm.error && elm.value)
    );
  }

  preparePayloadForFormSubmission() {
    let payload = {};
    this.state.textfieldsArr.forEach(e => (payload[e.name] = e.value));
    return payload;
  }

  checkAssetsAndLiabilities() {
    let assetsValue = this.state.textfieldsArr.find(
      e => e.name === "assetsValue"
    );
    let liabilitiesValue = this.state.textfieldsArr.find(
      e => e.name === "liabilitiesValue"
    );
    return assetsValue > liabilitiesValue ? true : false;
  }

  render() {
    return (
      <Container maxWidth="md">
        <h2>Credit card form (Made Manually)</h2>
        <form action="" className="credit__card__app__form">
          {this.state.textfieldsArr.map((elm, ind) => (
            <React.Fragment key={elm.name}>
              <input
                name={elm.name}
                type={elm.type}
                value={elm.value}
                placeholder={elm.label}
                onChange={this.handleOnChange.bind(this, ind)}
                onBlur={this.validateField.bind(this, ind)}
              />
              {elm.error ? (
                <small className="error__msg__helper">
                  {elm.errorMessages.map((msg, index) => (
                    <span key={msg}>
                      {index === elm.errorMessages.length - 1
                        ? msg
                        : `${msg}, `}
                    </span>
                  ))}
                </small>
              ) : null}
            </React.Fragment>
          ))}
          <div className="btn_container">
            <button key={10923} onClick={() => this.props.quitForm(null)}>
              Quit Application
            </button>
            <button key={93499} onClick={this.handleFormSubmit.bind(this)}>
              Submit
            </button>
          </div>
        </form>
      </Container>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onFormSubmit: payload =>
      dispatch({ type: "CREDITCARD__APP__FORM__SUBMIT", payload })
  };
};
export default connect(
  null,
  mapDispatchToProps
)(CreditCardAppForm);
