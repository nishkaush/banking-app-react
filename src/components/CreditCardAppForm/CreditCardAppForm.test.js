import React from "react";
import { shallow } from "enzyme";
import { CreditCardAppForm } from "./CreditCardAppForm";

describe("<CreditCardAppForm/>", () => {
  const wrapper = shallow(<CreditCardAppForm />);
  it("renders credit card app form correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // CHECKING THE FULL NAME FIELD
  describe("checks the Full Name Field", () => {
    it("auto focusses on the full name field first", () => {
      expect(wrapper.find("#fullName").prop("autoFocus")).toEqual(true);
    });
    it("updates the value of the input field, as user types", () => {
      wrapper
        .find("#fullName")
        .simulate("change", { target: { value: "success" } });
      expect(wrapper.find("#fullName").props().value).toEqual("success");
    });

    it("shows an error msg if the fullName isn't atleast 8 characters long", () => {
      wrapper
        .find("#fullName")
        .simulate("change", { target: { value: "error" } });
      wrapper
        .find("#fullName")
        .simulate("blur", { target: { value: "error" } });
      expect(
        wrapper
          .find(".error__msg__helper")
          .at(0)
          .contains(<span>Must be atleast 8 characters long</span>)
      ).toEqual(true);
    });

    it("shows 2 error messages if no fullName is provided", () => {
      wrapper.find("#fullName").simulate("change", { target: { value: "" } });
      wrapper.find("#fullName").simulate("blur", { target: { value: "" } });
      wrapper
        .find(".error__msg__helper")
        .at(0)
        .contains([
          <span>Full Name is required, </span>,
          <span>Must be atleast 8 characters long</span>
        ]);
    });

    // CHECKING THE EMAIL FIELD
    describe("checks the email field", () => {
      it("captures the user input as the user types the email", () => {
        wrapper
          .find("#email")
          .simulate("change", { target: { value: "success" } });
        expect(wrapper.find("#email").props().value).toEqual("success");
      });

      it("throws an error message if user enters an invalid email", () => {
        wrapper
          .find("#email")
          .simulate("change", { target: { value: "errw" } });
        wrapper.find("#email").simulate("blur", { target: { value: "errw" } });
        expect(
          wrapper
            .find(".error__msg__helper")
            .at(1)
            .childAt(0)
            .text()
        ).toEqual("Must be a valid email address");
      });

      it("throws 2 error messages when user doesn't input a value", () => {
        wrapper.find("#email").simulate("change", { target: { value: "" } });
        wrapper.find("#email").simulate("blur", { target: { value: "" } });

        // expect(
        //   wrapper
        //     .find(".error__msg__helper")
        //     .at(1)
        //     .text()
        // ).toEqual("Email is required, Must be a valid email address");
        // OR
        expect(
          wrapper
            .find(".error__msg__helper")
            .at(1)
            .contains([
              <span>Email is required, </span>,
              <span>Must be a valid email address</span>
            ])
        ).toEqual(true);
      });
    });

    // checking the FORM Submission
    describe("submits the form un-Successfully", () => {
      // step 1. mock ---> this.props.onShowDialogAlert = jest.fn()
      const onShowDialogAlert = jest.fn();

      // First Scenario :
      //all fieldsinvalid and assets < liabilities
      it("shows error for when not all required fields have value and when assets < liabilities", () => {
        let currentState = { ...wrapper.state() };
        currentState.textfieldsArr.forEach(e => (e.value = ""));
        wrapper.setState({ textfieldsArr: currentState.textfieldsArr });

        // set the props on thw wrapper
        wrapper.setProps({ onShowDialogAlert });

        // Try Submitting the form and check various scenrios
        wrapper.find("#submitBtn").simulate("click", {
          preventDefault: () => {}
        });

        // ---> check if mock fn called with specific args like error msg and status
        let msg =
          "Provide a valid value for all required fields (marked with *) and ensure assets value is bigger than liabilities";
        let status = "Error";

        expect(onShowDialogAlert).toHaveBeenLastCalledWith({ status, msg });
        expect(onShowDialogAlert).toHaveBeenCalledTimes(1);
      });

      it("shows the correct error message when some required fields are invalid and Assets>Liablities", () => {
        // Second scenario :
        // all fields Invalid but assets> liablities
        // ---> check if mock fn called with specific args like error msg and status
        let currentState = { ...wrapper.state() };
        let obj = { assetsValue: 150, liabilitiesValue: 100, email: "srsre" };
        currentState.textfieldsArr.forEach(e =>
          obj[e.name] ? (e.value = obj[e.name]) : null
        );
        wrapper.setState({ textfieldsArr: currentState.textfieldsArr });
        wrapper.find("#submitBtn").simulate("click", {
          preventDefault: () => {}
        });
        let msg =
          "Provide a valid value for all required fields (marked with *)";
        let status = "Error";
        expect(onShowDialogAlert).toHaveBeenLastCalledWith({ status, msg });
      });

      //Third scenario :
      //all fields are valid but assets < liabilities
      // ---> check if mock fn called with specific args like error msg and status
      it("shows error when all fields have a valid value but Assets < Liabilities", () => {
        let currentState = { ...wrapper.state() };
        let obj = {
          fullName: "Maximillian",
          email: "lol@lol.com",
          creditLimit: 50000,
          assetsValue: 90,
          liabilitiesValue: 200
        };
        currentState.textfieldsArr.forEach(e => {
          if (obj[e.name]) {
            e.value = obj[e.name];
            e.error = false;
          }
        });
        wrapper.setState({ textfieldsArr: currentState.textfieldsArr });
        wrapper.find("#submitBtn").simulate("click", {
          preventDefault: () => {}
        });
        let msg = "Make sure assets value is bigger than liabilities";
        let status = "Error";
        expect(onShowDialogAlert).toHaveBeenLastCalledWith({ status, msg });
      });
    });
  });
});
