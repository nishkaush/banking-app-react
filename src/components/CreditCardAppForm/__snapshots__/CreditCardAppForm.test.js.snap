// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreditCardAppForm/> checks the Full Name Field Successful form submission successfully submits the form 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="md"
>
  <h2>
    Credit card form (Made Manually)
  </h2>
  <form
    action=""
    className="credit__card__app__form"
  >
    <label
      htmlFor="fullName"
    >
      Full Name *
    </label>
    <input
      autoFocus={true}
      id="fullName"
      name="fullName"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="Maximillian"
    />
    <label
      htmlFor="email"
    >
      Email *
    </label>
    <input
      autoFocus={false}
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      type="email"
      value="lol@lol.com"
    />
    <label
      htmlFor="creditLimit"
    >
      Credit Limit ($) *
    </label>
    <input
      autoFocus={false}
      id="creditLimit"
      name="creditLimit"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value={50000}
    />
    <label
      htmlFor="assetsValue"
    >
      Total Assets Value ($) *
    </label>
    <input
      autoFocus={false}
      id="assetsValue"
      name="assetsValue"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value={90}
    />
    <label
      htmlFor="liabilitiesValue"
    >
      Total Liabilities Value ($) *
    </label>
    <input
      autoFocus={false}
      id="liabilitiesValue"
      name="liabilitiesValue"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value={200}
    />
    <div
      className="btn_container"
    >
      <button
        key="10923"
        onClick={[Function]}
      >
        Quit Application
      </button>
      <button
        id="submitBtn"
        key="93499"
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
  </form>
</WithStyles(ForwardRef(Container))>
`;

exports[`<CreditCardAppForm/> renders credit card app form correctly 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="md"
>
  <h2>
    Credit card form (Made Manually)
  </h2>
  <form
    action=""
    className="credit__card__app__form"
  >
    <label
      htmlFor="fullName"
    >
      Full Name *
    </label>
    <input
      autoFocus={true}
      id="fullName"
      name="fullName"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      htmlFor="email"
    >
      Email *
    </label>
    <input
      autoFocus={false}
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      type="email"
      value=""
    />
    <label
      htmlFor="creditLimit"
    >
      Credit Limit ($) *
    </label>
    <input
      autoFocus={false}
      id="creditLimit"
      name="creditLimit"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value=""
    />
    <label
      htmlFor="assetsValue"
    >
      Total Assets Value ($) *
    </label>
    <input
      autoFocus={false}
      id="assetsValue"
      name="assetsValue"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value=""
    />
    <label
      htmlFor="liabilitiesValue"
    >
      Total Liabilities Value ($) *
    </label>
    <input
      autoFocus={false}
      id="liabilitiesValue"
      name="liabilitiesValue"
      onBlur={[Function]}
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="btn_container"
    >
      <button
        key="10923"
        onClick={[Function]}
      >
        Quit Application
      </button>
      <button
        id="submitBtn"
        key="93499"
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
  </form>
</WithStyles(ForwardRef(Container))>
`;
